### Brief Notes on the History of NumPy

#### Predecessors: Numeric (1995), Numarray (2001) - limitations

#### NumPy Emergence (2005): Combined strengths, became standard

#### Significance: Foundation for data science libraries

#### Why It Was Needed: Python lists inefficient, predecessor limitations

---

## What Was Used Before NumPy?
* Python Lists: Versatile but inefficient for numerical operations
* Numeric & Numarray: Precursors with limitations

---

## Characteristics of NumPy Arrays
* Homogeneous Data: All elements same type
* Fixed Size: Size cannot be changed after creation
* Efficient Memory Storage: Contiguous memory blocks
* Multidimensional: 1D, 2D, 3D, etc.

### NumPy Data Types and Their Storage Usage

When creating a NumPy array, the `dtype` parameter allows you to specify the data type of the array elements. This affects both the **range of values** and the **amount of memory** each element consumes.

---

## **Integers (`int`)**

| Data Type | Size | Range                           |
|-----------|------|--------------------------------|
| `int8`    | 1 byte (8 bits)  | -128 to 127                  |
| `int16`   | 2 bytes (16 bits) | -32,768 to 32,767            |
| `int32`   | 4 bytes (32 bits) | -2.1 billion to 2.1 billion  |
| `int64`   | 8 bytes (64 bits) | Very large numbers           |

---

## **Floating Point (`float`)**

| Data Type  | Size | Range & Precision                 |
|------------|------|-----------------------------------|
| `float16`  | 2 bytes (16 bits) | Half precision (lower accuracy)  |
| `float32`  | 4 bytes (32 bits) | Single precision                 |
| `float64`  | 8 bytes (64 bits) | Double precision (default)       |

---

## **Boolean (`bool`)**

- **Data Type:** `bool_`  
- **Size:** **1 byte (8 bits)**  
- **Values:** `True` or `False`  


# ***ARRAYS***


2d array read as  rowXcolumn

3d array read as  layerXrowXcolumn

4d arrray read as layerXdepthXrowXcolumn

5d array read as  layersXblocksXdepthXrowsXcolumns

# 1D arrays
have only one dimension.

No Layers: Layers are a concept for higher-dimensional arrays (3D and above).

No Rows and Columns (in the traditional sense): While you can conceptually think of a 1D array as a single row or a single column, it's more accurately described as a sequence of elements.

In [ ]:
# Q *1*. create an 1D array A with the int 16 data type

import numpy as np

A = np.array ([1,2,3,4,5])  # here i am not mentioning the dtype
print('array A is :',A)
print(A.dtype) # lets see what kind of default type of data it took
#or

B = np.array ([1,2,3,4,5,6],dtype='int32')
print('array B is :',B)
print(B.dtype)

array A is : [1 2 3 4 5]
int64
array B is : [1 2 3 4 5 6]
int32


NOTES -

# *** 2D Array:***

 a 2D array does not have layers.

Represents data in a tabular format with rows and columns.

Think of it like a spreadsheet or a matrix.

Has two dimensions: rows and columns

In [ ]:
# Q2. Create different  2D arrays and find their  shape and dimensions

import numpy as np

array_2d = np.array([[1,2]])

print('2D array with two elements within a single row:', array_2d)
print('2D array shape:', array_2d.shape) #tells the shape
print('2D array dimensions:', array_2d.ndim) #how many dimensions in it
print('2D array  memory usage using itemsize in bytes :', array_2d.itemsize) # memory usage
print('2D array  memory usage using nbytes in bytes :', array_2d.nbytes) #memory usage

2D array with two elements within a single row: [[1 2]]
2D array shape: (1, 2)
2D array dimensions: 2
2D array  memory usage using itemsize in bytes : 8
2D array  memory usage using nbytes in bytes : 16


Why itemsize and nbytes Are Different

Because--

itemsize gives memory used by one element in the array.

nbytes gives memory used by the entire array (i.e., all elements).

In [ ]:
import numpy as np

array_2da = np.array([[1,2],[3,4]])

print('2D array with four elements across two rows:', array_2da)
print('2D array shape:', array_2da.shape)
print('2D array dimensions:', array_2da.ndim)

2D array with four elements across two rows: [[1 2]
 [3 4]]
2D array shape: (2, 2)
2D array dimensions: 2


In [ ]:
import numpy as np

array_2db = np.array([[1,2],[3,4],[5,6],[7,8]])

print('2D array with eight elements across four rows:', array_2db)
print('2D array shape:', array_2db.shape)
print('2D array dimensions:', array_2db.ndim)

2D array with eight elements across four rows: [[1 2]
 [3 4]
 [5 6]
 [7 8]]
2D array shape: (4, 2)
2D array dimensions: 2


In [ ]:
import numpy as np

array_2d1 = np.array([[1,2,3],[4,5,6]])

print('2D array with six elements across two rows:', array_2d1)
print('2D array shape:', array_2d1.shape)
print('2D array dimensions:', array_2d1.ndim)

2D array with six elements across two rows: [[1 2 3]
 [4 5 6]]
2D array shape: (2, 3)
2D array dimensions: 2


In [ ]:
import numpy as np

array_2d1a = np.array([[1,2,3],[4,5,6],[7,8,9]])

print('2D array with nine elements across three rows:', array_2d1a)
print('2D array shape:', array_2d1a.shape)
print('2D array dimensions:', array_2d1a.ndim)

2D array with nine elements across three rows: [[1 2 3]
 [4 5 6]
 [7 8 9]]
2D array shape: (3, 3)
2D array dimensions: 2


In [ ]:
import numpy as np

array_2d2 = np.array([[1,2,3,4],[5,6,7,8]])

print('2D array with eight elements across two rows:', array_2d2)
print('2D array shape:', array_2d2.shape)
print('2D array dimensions:', array_2d2.ndim)

2D array with eight elements across two rows: [[1 2 3 4]
 [5 6 7 8]]
2D array shape: (2, 4)
2D array dimensions: 2


In [ ]:
import numpy as np

array_2d2a = np.array([[1,2,3,4],[5,6,7,8],[9,0,1,2]])

print('2D array with twelve elements across three rows:', array_2d2a)
print('2D array shape:', array_2d2a.shape)
print('2D array dimensions:', array_2d2a.ndim)

2D array with twelve elements across three rows: [[1 2 3 4]
 [5 6 7 8]
 [9 0 1 2]]
2D array shape: (3, 4)
2D array dimensions: 2


## 3d arrays

shape will be read as layer,rows,columns

Max Layers: Limited by the memory available and the size of rows and columns.

#NOTES

1D + 2D = not 3D,

But multiple 2D stacked = 3D ✅

In [ ]:
#Q3 Lets create 3d arrays with different numbers of elements
# shape will be read as layer,rows,columns
# Max Layers: Limited by the memory available and the size of rows and columns.
import numpy as np

array_3d = np.array([[[1,2]]])
print('array_3d with two elements within a single layer and row:',array_3d)
print('array_3d shape:',array_3d.shape)
print('array_3d dimensions:',array_3d.ndim)


array_3d with two elements within a single layer and row: [[[1 2]]]
array_3d shape: (1, 1, 2)
array_3d dimensions: 3


In [ ]:
#3D array with one layer
import numpy as np

array_3d1 = np.array([[[1,2],[3,4]]])
print('this is what array_3d1 looks like :',array_3d1)
print('array_3d1 shape:',array_3d1.shape)
print('array_3d1 dimensions:',array_3d1.ndim)

this is what array_3d1 looks like : [[[1 2]
  [3 4]]]
array_3d1 shape: (1, 2, 2)
array_3d1 dimensions: 3


In [ ]:
#3d array with 2 layers
import numpy as np

array_3d_ = np.array([ [[1,2],[3,4]] , [[1,2],[3,4]] ])
print('this is what array_3d_ looks like :',array_3d_)
print('array_3d_ shape:',array_3d_.shape)
print('array_3d_ dimensions:',array_3d_.ndim)

this is what array_3d_ looks like : [[[1 2]
  [3 4]]

 [[1 2]
  [3 4]]]
array_3d_ shape: (2, 2, 2)
array_3d_ dimensions: 3


In [ ]:
import numpy as np

# Create a 3D array with 3 layers
array_3d = np.array([
    [[1, 2, 3], [4, 5, 6]],  # Layer 1
    [[7, 8, 9], [10, 11, 12]],  # Layer 2
    [[13, 14, 15], [16, 17, 18]]  # Layer 3
])

print("3D Array:")
print(array_3d)
print("Shape:", array_3d.shape)  # Output: (3, 2, 3)


3D Array:
[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]

 [[13 14 15]
  [16 17 18]]]
Shape: (3, 2, 3)


In [ ]:
import numpy as np

# Create a 3D array with 4 layers
array_3d = np.array([
    [[1, 2, 3], [4, 5, 6]],  # Layer 1
    [[7, 8, 9], [10, 11, 12]],  # Layer 2
    [[13, 14, 15], [16, 17, 18]],  # Layer 3
    [[19, 20, 21], [22, 23, 24]]  # Layer 4
])

print("3D Array:")
print(array_3d)
print("Shape:", array_3d.shape)  # Output: (4, 2, 3)

3D Array:
[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]

 [[13 14 15]
  [16 17 18]]

 [[19 20 21]
  [22 23 24]]]
Shape: (4, 2, 3)


# 4D arrays

Shape: (layers, depth, rows, columns)

Max Layers: Depends on depth, rows, and columns sizes.

In [ ]:
# lets create different types of 4D arrays with different number of elements
# Shape: (layers, depth, rows, columns)
# Max Layers: Depends on depth, rows, and columns sizes.
import numpy as np


array_4D=np.array([[[[1,2]]]])
print(array_4D)
print('array_4D shape:',array_4D.shape)
print('array_4D dimensions:',array_4D.ndim)

[[[[1 2]]]]
array_4D shape: (1, 1, 1, 2)
array_4D dimensions: 4


In [ ]:
# 4D array with one layer

import numpy as np

array_4D=np.array([[[[1,2],[2,3]]]])
print(array_4D)
print('array_4D shape:',array_4D.shape)
print('array_4D dimensions:',array_4D.ndim)

[[[[1 2]
   [2 3]]]]
array_4D shape: (1, 1, 2, 2)
array_4D dimensions: 4


In [ ]:
# 4D array with two layer

import numpy as np

# Create two 3D arrays (each representing a layer)
layer1 = np.array([ [[1, 2], [3, 4]], [[5, 6], [7, 8]] ])  # Shape: (2, 2, 2)
layer2 = np.array([[[9, 10], [11, 12]], [[13, 14], [15, 16]]])  # Shape: (2, 2, 2)

# Combine the layers into a 4D array
four_d_array = np.array([layer1, layer2])

print("4D Array:")
print(four_d_array)
print("Shape:", four_d_array.shape)  # Output: (2, 2, 2, 2)



4D Array:
[[[[ 1  2]
   [ 3  4]]

  [[ 5  6]
   [ 7  8]]]


 [[[ 9 10]
   [11 12]]

  [[13 14]
   [15 16]]]]
Shape: (2, 2, 2, 2)


In [ ]:
# 4D array with three layers

import numpy as np

# Create three 3D arrays (each representing a layer)
layer1 = np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])  # Shape: (2, 2, 2)
layer2 = np.array([[[9, 10], [11, 12]], [[13, 14], [15, 16]]])  # Shape: (2, 2, 2)
layer3 = np.array([[[17, 18], [19, 20]], [[21, 22], [23, 24]]])  # Shape: (2, 2, 2)

# Combine the layers into a 4D array
four_d_array = np.array([layer1, layer2, layer3])

print("4D Array:")
print(four_d_array)
print("Shape:", four_d_array.shape)  # Output: (3, 2, 2, 2)



4D Array:
[[[[ 1  2]
   [ 3  4]]

  [[ 5  6]
   [ 7  8]]]


 [[[ 9 10]
   [11 12]]

  [[13 14]
   [15 16]]]


 [[[17 18]
   [19 20]]

  [[21 22]
   [23 24]]]]
Shape: (3, 2, 2, 2)


In [ ]:
import numpy as np

# Create a 4D array with 4 layers, each layer being a 2x2x2 array
array_4d = np.array([
    [[[1, 2], [3, 4]], [[5, 6], [7, 8]]],  # Layer 1
    [[[9, 10], [11, 12]], [[13, 14], [15, 16]]],  # Layer 2
    [[[17, 18], [19, 20]], [[21, 22], [23, 24]]],  # Layer 3
    [[[25, 26], [27, 28]], [[29, 30], [31, 32]]]  # Layer 4
])

print("4D Array:")
print(array_4d)
print("Shape:", array_4d.shape)  # Output: (4, 2, 2, 2)


4D Array:
[[[[ 1  2]
   [ 3  4]]

  [[ 5  6]
   [ 7  8]]]


 [[[ 9 10]
   [11 12]]

  [[13 14]
   [15 16]]]


 [[[17 18]
   [19 20]]

  [[21 22]
   [23 24]]]


 [[[25 26]
   [27 28]]

  [[29 30]
   [31 32]]]]
Shape: (4, 2, 2, 2)


#####  ***NOTES FOR ARRAYS***
***5D Arrays***

**Shape:** (layers, blocks, depth, rows, columns)

**Total Elements:**

Total Elements = layers * blocks * depth * rows * columns

for more information please check the main lecture ***python_numpy.ipynb***

### Notes on Choosing Data Types (dtype)

When deciding on a data type (`dtype`), consider:

- **Range of values**:
  - Use `int8`, `int16`, or `int32` for integer data, depending on the range.
  - Use `float32` or `float64` for decimal (floating-point) data.
  
- **Precision required**:
  - If high precision is needed (e.g., scientific calculations), prefer `float64`.
  
- **Memory constraints**:
  - If working with very large datasets, choose smaller types like `int8` or `float32` to save memory.

# Slicing in NumPy Arrays (1D, 2D, 3D)

## General Slicing Syntax

`array[start:stop:step]`

| Parameter | Behavior |
|-----------|----------|
| `start` | Starting index (inclusive). Default: beginning |
| `stop` | Ending index (exclusive). Default: end |
| `step` | Elements to skip between selections. Default: 1 |

---

## General Rule to Avoid Confusion

### For 1D Arrays:
- Slicing operates on elements directly (no rows/columns)
- Example: `arr[start:stop:step]`

### For 2D Arrays:
- First index specifies rows, second specifies columns:
  - `a[start:stop:step, :]` → Row slicing
  - `a[:, start:stop:step]` → Column slicing

### For 3D Arrays:
- Follows similar pattern: `array[depth, row, column]`

---

## The `::` Operator Explained

When used without `start`/`stop`:
- Means "from beginning to end with this step size"

### Examples:
| Syntax | Behavior |
|--------|----------|
| `array[::2]` | Every other element |
| `array[1::3]` | Every 3rd element starting at index 1 |
| `array[::-1]` | Reverse the array |
| `a[0::2]` | From first row, every second element |

---


# 🧠 Understanding "Every Second Element" in NumPy Arrays

This note explains how to select **every second element** in:
- 1D array
- 2D array
- 3D array

---

## 📌 1D Array

```python

import numpy as np

a = np.array( [ 10, 20, 30, 40, 50, 60 ] )

print(a[ : : 2 ] )

output will be

[ 10 30 50 ]



## Practical 2D Example
```python

import numpy as np

arr = np.array ( [ [ 1 , 2 , 3 ] , [ 4 , 5 , 6 ] , [ 7 ,8 , 9 ] ] )

# Get every other row
print ( arr [ : : 2 , : ] )

# Get second column in reverse
print ( arr [ : ,  1 ] [ : : - 1 ] )

A small note: you could also write the second one as

 arr [ : : - 1 ,  1]

 to get the same result, which would reverse the rows first, then take the second column.

### **ARRAY 1D Exercise:**

```python
import numpy as np

a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14])

print(a)
```

## Questions

1. How can you access the first element (1) in the array?

2. What is the syntax to retrieve the last element (14) in the array?

3. How can you extract the elements 4, 5, and 6 from the array using slicing?

4. **What syntax would you use to select every third element in the array?**

5. **How would you reverse the array using slicing?**

6. What is the syntax to select the elements 7, 8, and 9 using slicing?

7. How can you replace the value 12 with 20 in the array?

8. How would you extract the first five elements from the array?

9. **What slicing syntax retrieves every second element starting from the second element?**

10. How would you replace the values 3, 4, and 5 with 30, 40, and 50 in the array?

In [ ]:
import numpy as np

a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14])
print(a)


print('Q1. first element:',a[0])
print('Q2.What is the syntax to retrieve the last element (14) in the array?:',a[-1])
print('Q3.extract the elements 4, 5, 6 from the array using slicing?:',a[3:6:1]) # here we are extracting multiple elements
print('Q4.What syntax would you use to select every third element in the array?:',a[::3]) #now I understand the concept  , here no rows no columns as it is 1D array just start:stop:stepsize
print('Q5.How would you reverse the array using slicing?:',a[::-1])
print('Q6.What is the syntax to select the elements 7, 8, 9 using slicing?:',a[6:9]) # or print (a[6:9:1])
#Q7 is in next tab
print('Q8.How would you extract the first five elements from the array?:',a[0:5:1])
print('Q9.What slicing syntax retrieves every second element starting from the second element?:',a[1::2])
#Q10 is in next tab

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14]
Q1. first element: 1
Q2.What is the syntax to retrieve the last element (14) in the array?: 14
Q3.extract the elements 4, 5, 6 from the array using slicing?: [4 5 6]
Q4.What syntax would you use to select every third element in the array?: [ 1  4  7 10 13]
Q5.How would you reverse the array using slicing?: [14 13 12 11 10  9  8  7  6  5  4  3  2  1]
Q6.What is the syntax to select the elements 7, 8, 9 using slicing?: [7 8 9]
Q8.How would you extract the first five elements from the array?: [1 2 3 4 5]
Q9.What slicing syntax retrieves every second element starting from the second element?: [ 2  4  6  8 10 12 14]


In [ ]:
import numpy as np

a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14])
print(a)
# Q7.How can you replace the value 12 with 20 in the array?
a[11]=20
print('new array:',a)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14]
new array: [ 1  2  3  4  5  6  7  8  9 10 11 20 13 14]


In [ ]:
#10. How would you replace the values 3, 4, 5 with 30, 40, 50 in the array?
import numpy as np

a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14])
print(a)

a[2:5] = [30, 40, 50]
print('Q10. Replace 3, 4, 5 with 30, 40, 50:', a)



[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14]
Q10. Replace 3, 4, 5 with 30, 40, 50: [ 1  2 30 40 50  6  7  8  9 10 11 12 13 14]


#**Exercise:2D ARRAY**

Notes - A 2D array can have just two layers


### ✅ Finding Specific Elements, Rows or Columns in NumPy Array

```python
import numpy as np

a = np.array([[1, 2, 3, 4, 5, 6, 7],
              [8, 9, 10, 11, 12, 13, 14]])
print(a)

**NumPy Array Indexing and Slicing Questions**

1. How do you access the first element (1) of the first row in a NumPy array?

2. What is the syntax to select the element 10 from the given array?

3. How can the entire first row be retrieved from the array?

4. What method would you employ to extract the third column from the array?

5. **How can every second element be selected from the first row of the array?**

6. How would you access the last element (14) of the second row?

7. **What syntax would you use to slice and retrieve elements from the second row, starting from the second column to the fifth column (inclusive)?**

8. **How would you find the indices of the values 3 and 10 in the third column of the array?**

9. How can the value 13 be replaced with 20 in the array `a`?

10. What syntax would you use to replace the values 3 and 10 in the third column with 5?

11. How would you replace the values in the third column with 50 and 60?

In [ ]:
import  numpy as np

a = np.array([[1,2, 3, 4, 5, 6, 7],[8, 9, 10, 11, 12, 13, 14]])
print(a)

print('Q1.access the first element (1) of the first row in a NumPy array?:',a[0,0]) # syntax is (row,column)with zero index

print('Q2.What is the syntax to select the element 10 from the given array?:',a[1,2])  # again syntax is (row,column)with zero index

print('Q3.How can the entire first row be retrieved from the array?:',a[0,::]) # (row index, column index(start:stop:size))

print('Q4.What method would you employ to extract the third column from the array?:',a[::,2])   # (row index(start:stop:size), column index)

print('Q5.How can every second element be selected from the first row of the array?:',a[0,::2])

print('Q6.How would you access the last element (14) of the second row?:',a[-1,-1])

print('Q7.What syntax would you use to slice and retrieve elements from the second row, starting from the second column to the fifth column (inclusive)?:',a[1,1:5:1])

print('Q8.How would you find the indices of the values 3 and 10 in the third column of the array?:',a[::,2])

#Q9 is below
#Q10 is below
#Q11 is below


[[ 1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14]]
Q1.access the first element (1) of the first row in a NumPy array?: 1
Q2.What is the syntax to select the element 10 from the given array?: 10
Q3.How can the entire first row be retrieved from the array?: [1 2 3 4 5 6 7]
Q4.What method would you employ to extract the third column from the array?: [ 3 10]
Q5.How can every second element be selected from the first row of the array?: [1 3 5 7]
Q6.How would you access the last element (14) of the second row?: 14
Q7.What syntax would you use to slice and retrieve elements from the second row, starting from the second column to the fifth column (inclusive)?: [ 9 10 11 12]
Q8.How would you find the indices of the values 3 and 10 in the third column of the array?: [ 3 10]


In [ ]:
# Q9.How can the value 13 be replaced with 20 in the array a?
import  numpy as np

a = np.array([[1,2, 3, 4, 5, 6, 7],[8, 9, 10, 11, 12, 13, 14]])
print ('old array :',a)

a[1,5]=20
print ('new array :',a)

# Q10.What syntax would you use to replace the values 3 and 10 in the third column with 5?  # important one
a[:,2] =[5,5]

print ('new array with 5,5 :',a)

#Q.11 How would you replace the values in the third column with 50 and 60? # important one

a[:,2]=[50,60]
print ('new array with 50,60 :',a)


old array : [[ 1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14]]
new array : [[ 1  2  3  4  5  6  7]
 [ 8  9 10 11 12 20 14]]
new array with 5,5 : [[ 1  2  5  4  5  6  7]
 [ 8  9  5 11 12 20 14]]
new array with 50,60 : [[ 1  2 50  4  5  6  7]
 [ 8  9 60 11 12 20 14]]


### Understanding a 3D NumPy Array

A 3D array can be thought of as a stack of 2D arrays (or layers). The axes are:

* **Axis 0 (Layers):** Refers to the layer index (e.g., layer1, layer2, layer3).
* **Axis 1 (Rows):** Refers to the rows within a layer.
* **Axis 2 (Columns):** Refers to the columns within a row.

**Basic Indexing in 3D Arrays**

The general syntax is:

    ```python
    array_3d[layer_index, row_index, column_index]

In [ ]:
# here I will create 3d array with possible 2,3,4 layers and then I will try how index and slicing works within 3d array

In [ ]:
# layer1 =np.array([1,2,3,4],[4,5,6,7]),
# layer2 =np.array([8,9,10,11][12,13,14,15]),
# layer3=np.array([16,17,18,19],[20,21,22,23])+

# I was thinkning them as  a 1D array but I was wrong , 1D array element goes continous in just [],
#  for using [][] , it has to be 2D array for example ([ [],[] ])

# **EXCERSIZE 3D NumPy Array**

In [ ]:
import numpy as np
layer1 = np.array([[1, 2, 3, 4], [4, 5, 6, 7]])
layer2 = np.array([[8, 9, 10, 11], [12, 13, 14, 15]])
layer3 = np.array([[16, 17, 18, 19], [20, 21, 22, 23]])

array_3d_3layer = np.array([layer1,layer2,layer3]) # here I was missing []
print(array_3d_3layer)

print('shape is (layer,rows,columns) ',array_3d_3layer.shape)
print('number of dimensions are ',array_3d_3layer.ndim)
print('bytes size of this array is ',array_3d_3layer.nbytes)
print('itemsize of the array is ',array_3d_3layer.itemsize)
print('data type  of the array is ',array_3d_3layer.dtype)

[[[ 1  2  3  4]
  [ 4  5  6  7]]

 [[ 8  9 10 11]
  [12 13 14 15]]

 [[16 17 18 19]
  [20 21 22 23]]]
shape is (layer,rows,columns)  (3, 2, 4)
number of dimensions are  3
bytes size of this array is  192
itemsize of the array is  8
data type  of the array is  int64


1. How would you access the first element (1) of the first layer (layer1)?

2. What is the syntax to retrieve the value 14 from the second layer (layer2)?

3. **How can you extract the entire second row from the third layer (layer3)?**

4. **What syntax would you use to retrieve all elements of the first column across all three layers?**

5. **How can you select all elements of the second layer?**

6. How would you extract the values [3, 10, 18] from the third column across all layers?

7. **What syntax retrieves the sub-array consisting of the first two rows and first two columns of the second layer (layer2)?>8,9,12,13 **

8. How can you replace the value 9 in the second layer (layer2) with 90?

9. **What is the syntax to select every second element in the first row of all three layers?** 1,3,8,10,16,18

10. How would you replace the entire second row of layer3 with [50, 60, 70, 80]?

In [ ]:
print('answer 1 is:',array_3d_3layer[0,0,0])
print('answer 2 is:',array_3d_3layer[1,1,2])
print('answer 3 is:',array_3d_3layer[2,1:,:])
print('answer 4 is:',array_3d_3layer[:,:,0])
print('answer 5 is:',array_3d_3layer[1,:,:])

answer 1 is: 1
answer 2 is: 14
answer 3 is: [[20 21 22 23]]
answer 4 is: [[ 1  4]
 [ 8 12]
 [16 20]]
answer 5 is: [[ 8  9 10 11]
 [12 13 14 15]]


In [ ]:
print('answer 6 is:',array_3d_3layer[:, 0, 2])
print('answer 7 is:',array_3d_3layer[1,0:2,0:2])

answer 6 is: [ 3 10 18]
answer 7 is: [[ 8  9]
 [12 13]]


In [ ]:
# answer 8 is
array_3d_3layer[1,0,1] = 90
array_3d_3layer

array([[[ 1,  2,  3,  4],
        [ 4,  5,  6,  7]],

       [[ 8, 90, 10, 11],
        [12, 13, 14, 15]],

       [[16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [ ]:
# answer 9 is
array_3d_3layer[:, 0, ::2]

array([[ 1,  3],
       [ 8, 10],
       [16, 18]])

In [ ]:
#answer 10 is
array_3d_3layer[2,1,:]=(50,60,70,80)
array_3d_3layer

# ones matrix  and non one matrix with different values

In [ ]:
a_ones=np.ones([4,4])
a_ones

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [ ]:
# 1D Array Filled with 7
A1D=np.full([5,1],8) #with one column  # [ shapefill,value we want to see ]
A1D

array([[8],
       [8],
       [8],
       [8],
       [8]])

In [ ]:
A1D=np.full([5,],8) # with no column
A1D

array([8, 8, 8, 8, 8])

In [ ]:
a_full=np.full([1,2,3],3)  # here it will be a 3d array having value of 3 each
a_full

array([[[3, 3, 3],
        [3, 3, 3]]])

In [ ]:
a_full.shape

(1, 2, 3)

In [ ]:
import numpy as np
np.zeros((2,3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [ ]:
import numpy as np
np.arange(0,11,2)

array([ 0,  2,  4,  6,  8, 10])

In [ ]:
import numpy as np

np.zeros(3)

array([0., 0., 0.])

In [ ]:
# it will create matrices betweeen 0 and 5 having spaces of 10
import numpy as np
np.linspace(0,5,10)

array([0.        , 0.55555556, 1.11111111, 1.66666667, 2.22222222,
       2.77777778, 3.33333333, 3.88888889, 4.44444444, 5.        ])

# RANDOM METHOD

In [ ]:
import numpy as np

# Random 1D array
matrix = np.random.rand(5)
matrix

array([0.04997832, 0.56863686, 0.70488837, 0.16765612, 0.21705159])

In [ ]:
import numpy as np

# Random 2D matrix (2 rows, 3 columns)  # data type will be float
matrix = np.random.rand(2, 3)
matrix

array([[0.33566931, 0.2343839 , 0.73308921],
       [0.21515517, 0.92314554, 0.98560175]])

In [ ]:
# Random 3D matrix (2layers,3 rows, 3 columns) # data type will be float
matrix = np.random.rand(2, 3,3)
matrix

array([[[0.85323455, 0.60284607, 0.48156155],
        [0.02440374, 0.14909811, 0.83596825],
        [0.41361887, 0.27678307, 0.91630288]],

       [[0.52244965, 0.14454772, 0.49840938],
        [0.89942789, 0.21868104, 0.12095185],
        [0.81891799, 0.99612733, 0.84149822]]])

In [ ]:
matrix= np.random.randint(1,100,10)
matrix

array([85, 23,  1, 90, 75, 13, 77, 58,  9,  5])

In [ ]:
# Random integers between 10 and 100, size 3x4
matrix= np.random.randint(10, 100, size=(3, 4))
matrix

array([[89, 28, 79, 53],
       [92, 21, 96, 83],
       [93, 81, 34, 67]])

In [ ]:
# Random integers between 10 and 100, size 3x4
matrix= np.random.randint(10, 100, size=(3, 3,4))
matrix

array([[[12, 49, 76, 93],
        [95, 68, 14, 15],
        [72, 66, 63, 54]],

       [[25, 13, 92, 91],
        [80, 27, 70, 97],
        [92, 21, 66, 38]],

       [[86, 55, 55, 96],
        [30, 72, 22, 79],
        [89, 12, 71, 96]]])

In [ ]:
# 3. np.random.randn(): Random floats from normal distribution (mean=0, std=1)

matrix=np.random.randn(2, 3)
matrix

array([[-0.99473059, -1.18221611,  0.80694955],
       [-1.55257429, -1.92691177, -0.22883132]])

In [ ]:
matrix=np.random.randn(5, 4)
matrix

array([[-0.56656061, -0.25603957, -0.03375182,  2.0613414 ],
       [ 0.74853795,  0.02474047,  0.85714071, -1.05697548],
       [-0.00469889,  0.21510955, -0.80519199, -1.01548274],
       [ 0.73713243,  0.47689455,  0.25147011, -0.21033267],
       [ 0.02050112,  0.91254876, -0.77254604, -0.68529425]])

#### following notes --❌ You can't get true integers that follow a normal distribution (mean 0, std 1) because the normal distribution is continuous — it produces float values, not integers.

# 🧪 NumPy Random Seed — Notes

## What is `np.random.seed()`?

- `np.random.seed()` sets the **starting point** for generating random numbers.
- It ensures **reproducibility**, meaning you'll get the **same random numbers** every time you run the code.

---

## Why Use a Seed?

- Random results are usually different each time.
- A seed "locks" the randomness so:
  - You can debug code consistently.
  - Others running your code get the same output.
  - Experiments in ML become reproducible.

---

## Syntax

```python
import numpy as np

np.random.seed(101)  # 101 is just an example; can be any integer


In [ ]:
import numpy as np

np.random.seed(101)
print(np.random.randint(1, 100, 3))  # Output: [96 13 73]

np.random.seed(101)
print(np.random.randint(1, 100, 3))  # Output: [96 13 73] again!


[96 12 82]
[96 12 82]


In [ ]:
import numpy as np

np.random.seed(101)
print(np.random.randint(1, 100, 3))

np.random.seed(202)
print(np.random.randint(1, 100, 5))  # Always prints [90 45 53 25 29]


[96 12 82]
[20 24 61 77 41]


In [ ]:
#  Simulate Structured Data with Columns (Fake Dataset)

import pandas as pd

data = {
    "age": np.random.randint(18, 60, 100),
    "income": np.random.randint(30000, 100000, 100),
    "score": np.random.rand(100) * 100
}

df = pd.DataFrame(data)
print(df.head())


   age  income      score
0   30   60945  77.669793
1   44   80970  58.215760
2   19   54044  79.772986
3   20   91455  49.847218
4   36   51055  60.609853


In [ ]:
df.shape

(100, 3)

In [ ]:
data.shape

(100, 3)

In [ ]:
before=np.array([[1,2,3,4],[5,6,7,8]])
print(before.shape)

(2, 4)


In [ ]:
after =before.reshape((8,1))
print(after)

(2, 4)
[[1]
 [2]
 [3]
 [4]
 [5]
 [6]
 [7]
 [8]]


array([ 0,  2,  4,  6,  8, 10])

array([[0., 0., 0.],
       [0., 0., 0.]])

# Axis mean in Python

Think of axis as the direction in which an operation (like sum, mean, etc.) is performed.

axis=0 → down the rows, operate column-wise

axis=1 → across the columns, operate row-wise



In [2]:
#example
import numpy as np
data = np.array([
    [10, 20, 30],
    [40, 50, 60],
    [70, 80, 90]
])

print(data)

[[10 20 30]
 [40 50 60]
 [70 80 90]]


In [3]:
data.sum(axis=0)

array([120, 150, 180])

What happens?

Column 1 → 10 + 40 + 70 = 120

Column 2 → 20 + 50 + 80 = 150

Column 3 → 30 + 60 + 90 = 180

In [4]:
data.sum(axis=1)

array([ 60, 150, 240])

What happens?

Row 1 → 10 + 20 + 30 = 60

Row 2 → 40 + 50 + 60 = 150

Row 3 → 70 + 80 + 90 = 240

# NumPy Attributes and Methods by Topic

## 1. Array Properties (Attributes)

| Attribute | Example | Explanation |
|-----------|---------|-------------|
| `shape` | `arr.shape` → `(3, 4)` | Returns dimensions of array as tuple |
| `size` | `arr.size` → `12` | Total number of elements in array |
| `ndim` | `arr.ndim` → `2` | Number of dimensions (axes) |
| `dtype` | `arr.dtype` → `int64` | Data type of array elements |
| `itemsize` | `arr.itemsize` → `8` | Size in bytes of each element |
| `nbytes` | `arr.nbytes` → `96` | Total bytes consumed by array |
| `T` | `arr.T` | Transpose of the array |

```python
arr = np.array([[1, 2, 3], [4, 5, 6]])
print(arr.shape)  # (2, 3)
```

---

## 2. Array Creation (Methods)

| Method | Example | Explanation |
|--------|---------|-------------|
| `np.array()` | `np.array([1, 2, 3])` | Create array from list/tuple |
| `np.zeros()` | `np.zeros((3, 4))` | Array filled with zeros |
| `np.ones()` | `np.ones((2, 3))` | Array filled with ones |
| `np.arange()` | `np.arange(0, 10, 2)` | Array with evenly spaced values |
| `np.linspace()` | `np.linspace(0, 1, 5)` | Array with specified number of points |
| `np.eye()` | `np.eye(3)` | Identity matrix |
| `np.random.rand()` | `np.random.rand(3, 3)` | Random values in [0, 1) |

```python
arr = np.arange(0, 10, 2)  # [0, 2, 4, 6, 8]
```

---

## 3. Array Reshaping (Methods)

| Method | Example | Explanation |
|--------|---------|-------------|
| `reshape()` | `arr.reshape(3, 4)` | Change shape without changing data |
| `flatten()` | `arr.flatten()` | Convert to 1D array (copy) |
| `ravel()` | `arr.ravel()` | Convert to 1D array (view if possible) |
| `transpose()` | `arr.transpose()` | Permute dimensions |
| `resize()` | `arr.resize((3, 4))` | Change shape in-place |

```python
arr = np.array([[1, 2], [3, 4], [5, 6]])
new = arr.reshape(2, 3)  # [[1, 2, 3], [4, 5, 6]]
```

---

## 4. Mathematical Operations (Methods)

| Method | Example | Explanation |
|--------|---------|-------------|
| `sum()` | `arr.sum()` | Sum of all elements |
| `mean()` | `arr.mean()` | Average of elements |
| `std()` | `arr.std()` | Standard deviation |
| `min()` | `arr.min()` | Minimum value |
| `max()` | `arr.max()` | Maximum value |
| `cumsum()` | `arr.cumsum()` | Cumulative sum |
| `prod()` | `arr.prod()` | Product of all elements |

```python
arr = np.array([1, 2, 3, 4, 5])
print(arr.mean())  # 3.0
```

---

## 5. Indexing and Slicing (Methods)

| Method | Example | Explanation |
|--------|---------|-------------|
| `take()` | `arr.take([0, 2])` | Select elements by indices |
| `compress()` | `arr.compress([True, False, True])` | Select by boolean mask |
| `argmax()` | `arr.argmax()` | Index of maximum value |
| `argmin()` | `arr.argmin()` | Index of minimum value |
| `nonzero()` | `arr.nonzero()` | Indices of non-zero elements |

```python
arr = np.array([10, 20, 30, 40])
idx = arr.argmax()  # 3 (index of 40)
```

---

## 6. Array Manipulation (Methods)

| Method | Example | Explanation |
|--------|---------|-------------|
| `append()` | `np.append(arr, [4, 5])` | Append values to end |
| `insert()` | `np.insert(arr, 1, 99)` | Insert values at index |
| `delete()` | `np.delete(arr, 2)` | Delete elements at index |
| `concatenate()` | `np.concatenate([a, b])` | Join arrays along axis |
| `split()` | `np.split(arr, 3)` | Split array into multiple sub-arrays |
| `sort()` | `arr.sort()` | Sort array in-place |

```python
arr = np.array([3, 1, 4, 2])
arr.sort()  # [1, 2, 3, 4]
```

---

## 7. Statistical Operations (Methods)

| Method | Example | Explanation |
|--------|---------|-------------|
| `var()` | `arr.var()` | Variance of elements |
| `median()` | `np.median(arr)` | Median value |
| `percentile()` | `np.percentile(arr, 50)` | nth percentile |
| `corrcoef()` | `np.corrcoef(x, y)` | Correlation coefficient |
| `cov()` | `np.cov(arr)` | Covariance matrix |

```python
arr = np.array([1, 2, 3, 4, 5])
variance = arr.var()  # 2.0
```

---

## 8. Linear Algebra (Methods)

| Method | Example | Explanation |
|--------|---------|-------------|
| `dot()` | `arr.dot(other)` | Dot product of arrays |
| `np.linalg.inv()` | `np.linalg.inv(matrix)` | Matrix inverse |
| `np.linalg.det()` | `np.linalg.det(matrix)` | Determinant |
| `np.linalg.eig()` | `np.linalg.eig(matrix)` | Eigenvalues and eigenvectors |
| `np.matmul()` | `np.matmul(a, b)` | Matrix multiplication |

```python
a = np.array([[1, 2], [3, 4]])
det = np.linalg.det(a)  # -2.0
```

---

## 9. Type Conversion (Methods)

| Method | Example | Explanation |
|--------|---------|-------------|
| `astype()` | `arr.astype(float)` | Convert to different data type |
| `tolist()` | `arr.tolist()` | Convert to Python list |
| `copy()` | `arr.copy()` | Create deep copy of array |
| `view()` | `arr.view()` | Create view (shallow copy) |

```python
arr = np.array([1, 2, 3])
float_arr = arr.astype(float)  # [1.0, 2.0, 3.0]
```

---

## 10. Axis Operations (Methods with axis parameter)

| Method | Example | Explanation |
|--------|---------|-------------|
| `sum(axis=)` | `arr.sum(axis=0)` | Sum along specified axis |
| `mean(axis=)` | `arr.mean(axis=1)` | Mean along specified axis |
| `max(axis=)` | `arr.max(axis=0)` | Maximum along axis |
| `argmax(axis=)` | `arr.argmax(axis=1)` | Index of max along axis |

```python
arr = np.array([[1, 2], [3, 4]])
col_sum = arr.sum(axis=0)  # [4, 6] (sum each column)
```

---

## Quick Reference: Attributes vs Methods

**Attributes** (no parentheses): Properties of the array that you access directly
- `arr.shape`, `arr.size`, `arr.dtype`, `arr.ndim`, `arr.T`

**Methods** (with parentheses): Actions you perform on the array
- `arr.sum()`, `arr.reshape()`, `arr.mean()`, `arr.sort()`